ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs

```diff
@@ -0,0 +1,80 @@
+
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Data.List (intercalate)
+import           Ouroboros.Consensus.Block.Abstract (HeaderHash)
+import           Ouroboros.Consensus.Util.Condense (condense)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           System.Random.Stateful (runSTGen_)
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.Genesis.Setup.Classifiers
+import           Test.Consensus.PeerSimulator.Run (noTimeoutsSchedulerConfig, SchedulerConfig (scEnableGdd))
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PointSchedule
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck
+import           Test.QuickCheck.Extras (unsafeMapSuchThatJust)
+import           Test.QuickCheck.Random (QCGen)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestBlock, unTestHash)
+
+tests :: TestTree
+tests =
+  testProperty "uniform random schedule" prop_uniformSchedule
+
+genChainsAndSchedule :: PointScheduleConfig -> ScheduleType -> QC.Gen (GenesisTest, PointSchedule)
+genChainsAndSchedule scheduleConfig scheduleType =
+  unsafeMapSuchThatJust do
+    numAdversaries <- QC.choose (1, 4)
+    gt <- genChains numAdversaries
+    seed :: QCGen <- arbitrary
+    pure ((gt,) <$> runSTGen_ seed (\ g -> genSchedule g scheduleConfig scheduleType (gtBlockTree gt)))
+
+prop_uniformSchedule :: QC.Gen QC.Property
+prop_uniformSchedule = do
+  (genesisTest, schedule) <- genChainsAndSchedule scheduleConfig Uniform
+  let Classifiers {..} = classifiers genesisTest
+
+  pure $ withMaxSuccess 10 $
+    classify genesisWindowAfterIntersection "Full genesis window after intersection" $
+    runSimOrThrow $
+      runTest
+        (noTimeoutsSchedulerConfig scheduleConfig) {scEnableGdd = True}
+        genesisTest
+        schedule
+        $ exceptionCounterexample $ \StateView{svSelectedChain} killed ->
+            classify (immutableTipIsOrigin svSelectedChain) "Immutable tip is Origin" $
+            classify (length killed > 0) "Some peers were killed" $
+            classify (length killed == length (peerIds schedule) - 1) "All adversaries were killed" $
+            killCounterexample killed $
+            immutableTipIsHonest svSelectedChain
+
+  where
+    killCounterexample = \case
+      [] -> id
+      killed -> counterexample ("Some peers were killed: " ++ intercalate ", " (condense <$> killed))
+
+    immutableTipIsOrigin :: TestFragH -> Bool
+    immutableTipIsOrigin frag = case AF.anchor frag of
+        AF.AnchorGenesis -> True
+        AF.Anchor _ _ _  -> False
+
+    immutableTipIsHonest :: TestFragH -> Property
+    immutableTipIsHonest frag = case AF.anchor frag of
+        AF.AnchorGenesis   -> property True
+        AF.Anchor _ hash _ -> property (isHonestTestHeaderHash hash)
```

Comment:
Can't this be just a `TestFragH -> Bool` and everything works above because of the right typeclass instance for `Bool`?